File: src/com/zrss/leetcode/AddTwoNumbers.java
package com.zrss.leetcode;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode i1 = l1;
        ListNode i2 = l2;
        
        ListNode rel = new ListNode(-1);
        ListNode iR = rel;
        
        int carry = 0;
        
        while (i1 != null && i2 != null) {
            int val = i1.val + i2.val;
            val += carry;
            carry = val / 10;
            ListNode digit = new ListNode(val % 10);
            iR.next = digit;
            iR = digit;
            
            i1 = i1.next;
            i2 = i2.next;
        }
        
        ListNode remain = (i1 != null) ? i1 : i2;
        
        while (remain != null) {
            int val = remain.val + carry;
            carry = val / 10;
            ListNode digit = new ListNode(val % 10);
            iR.next = digit;
            iR = digit;
            
            remain = remain.next;
        }
        
        if (carry > 0) {
            ListNode digit = new ListNode(carry);
            iR.next = digit;
            iR = digit;
        }
        
        iR.next = null;
        
        return rel.next;
    }
    
    public static void main(String[] args) {
        ListNode l1 = new ListNode(9);
        l1.next = new ListNode(9);
        l1.next.next = new ListNode(9);
        
        ListNode l2 = new ListNode(9);
        l2.next = new ListNode(9);
        l2.next.next = new ListNode(9);
        
        AddTwoNumbers addTwoNumbers = new AddTwoNumbers();
        
        ListNode rel = addTwoNumbers.addTwoNumbers(l1, l2);
        
        for (ListNode itr = rel; itr != null; itr = itr.next) {
            System.out.println(itr.val);
        }
    }
}

File: AddTwoNumbers.cpp
#include <iostream>
#include <cstdio>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 模拟题
// 使用虚拟头结点可以简化代码
// 注意移动新链表的当前节点

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode vHead(0);
        ListNode* cur = &vHead;

        int in = 0;
        while (l1 != NULL && l2 != NULL) {
        	int tmp = in + l1->val + l2->val;
        	in = tmp / 10;
        	l1->val = tmp % 10;
        	cur->next = l1;
        	cur = l1;
        	l1 = l1->next;
        	l2 = l2->next;
        }

        // 直接判断剩余
    	ListNode* remain = l1 == NULL ? l2 : l1;
    	while (remain != NULL) {
    		int tmp = in + remain->val;
    		in = tmp / 10;
    		remain->val = tmp % 10;
    		cur->next = remain;
    		cur = remain;
    		remain = remain->next;
    	}

    	// 最高位有进位
    	if (in != 0) {
    		cur->next = new ListNode(in);
    		cur = cur->next;
    	}

        cur->next = NULL;
    	return vHead.next;
    }
};

int main(int argc, char const *argv[]) {
	Solution solution;
	return 0;
}